internal/handlers/common/update.go

```diff
@@ -18,33 +18,60 @@ import (
 	"fmt"
 
 	"github.com/FerretDB/FerretDB/internal/types"
-	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 	"github.com/FerretDB/FerretDB/internal/util/must"
 )
 
 // UpdateDocument updates the given document with a series of update operators.
-func UpdateDocument(doc, update *types.Document) error {
+// Returns true if document was changed.
+func UpdateDocument(doc, update *types.Document) (bool, error) {
 	for _, updateOp := range update.Keys() {
 		updateV := must.NotFail(update.Get(updateOp))
 
 		switch updateOp {
 		case "$set":
-			setDoc, err := AssertType[*types.Document](updateV)
-			if err != nil {
-				return err
+			msgFmt := `Modifiers operate on fields but we found type %s instead. ` +
+				`For example: {$mod: {<field>: ...}} not {$set: %s}`
+
+			if updateV == nil ||
+				updateV == types.Null {
+				return false, NewWriteErrorMsg(ErrFailedToParse, fmt.Sprintf(msgFmt, "null", "null"))
 			}
 
-			for _, setKey := range setDoc.Keys() {
-				setValue := must.NotFail(setDoc.Get(setKey))
-				if err = doc.Set(setKey, setValue); err != nil {
-					return lazyerrors.Error(err)
+			switch setDoc := updateV.(type) {
+			case *types.Document:
+				if setDoc.Len() == 0 {
+					return false, nil
+				}
+				var err error
+				for _, setKey := range setDoc.Keys() {
+					setValue := must.NotFail(setDoc.Get(setKey))
+					if err = doc.Set(setKey, setValue); err != nil {
+						return false, err
+					}
 				}
+				return true, nil
+
+			case *types.Array:
+				return false, NewWriteErrorMsg(ErrFailedToParse, fmt.Sprintf(msgFmt, "array", "[]"))
+
+			case float64:
+				return false, NewWriteErrorMsg(
+					ErrFailedToParse,
+					fmt.Sprintf(`Modifiers operate on fields but we found type double instead. `+
```

Comment:
You could use `AliasFromType` to get the correct type name for all supported types.

Reply:
It's shorter but it doesn't work...

Reply:
Maybe writing another function for type formatting in errors would help.

Reply:
alternatively, we can change the error msg for all types in order to not to marshal everything - I remember someone was about to create `.String()` method for the document only for the error message and we ended up with an alt message. 

I'll do it as we did before.